### PR TITLE
[BUGFIX] Scope `PandasDatasource` reader method types and serialization (1 of 2)

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-1735
- Allow `add_pandas` to leverage `read_<DATA_ASSET_TYPE>` methods by attaching `_data_context` in cases other than `pandas_default`
- Handle case where pandas `read_html` returns `list[pd.DataFrame]`
- Exclude fields of type `sqlalchemy.engine.Engine` or `sqlite3.Connection` from serialization
- Fix bug in `convert_to_json_serializable` that caused all non-convertable objects to become `None` instead of raising an error as expected

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/core/util.py [hunk 1]
```diff
@@ -401,15 +403,12 @@ def convert_to_json_serializable(  # noqa: C901 - complexity 32
...
-        if isinstance(data, StructType):
-            return dict(data.jsonValue())
+    if StructType and isinstance(data, StructType):
+        return dict(data.jsonValue())
 
-    else:
-        raise TypeError(
-            f"{str(data)} is of type {type(data).__name__} which cannot be serialized."
-        )
-    return None
```
Comment (NathanFarmer) [inline great_expectations/core/util.py:412]: This bug was a big deal! Anything that couldn't be serialized was ending up as `None` instead if raising the `TypeError` on line 409 when `StructType` was imported.

File: great_expectations/core/util.py [hunk 2]
```diff
@@ -401,15 +403,12 @@ def convert_to_json_serializable(  # noqa: C901 - complexity 32
...
-        if isinstance(data, StructType):
-            return dict(data.jsonValue())
+    if StructType is not None and isinstance(data, StructType):
+        return dict(data.jsonValue())
 
-    else:
-        raise TypeError(
-            f"{str(data)} is of type {type(data).__name__} which cannot be serialized."
-        )
-    return None
```
Comment (NathanFarmer) [inline great_expectations/core/util.py:412]: This bug was a big deal! Anything that couldn't be serialized was ending up as `None` instead if raising the `TypeError` on line 409 when `StructType` was imported.

File: great_expectations/datasource/fluent/sources.py [hunk 3]
```diff
@@ -284,6 +276,14 @@ def __getattr__(self, attr_name: str):
...
             def wrapped(name: str, **kwargs):
                 datasource = ds_constructor(name=name, **kwargs)
+                # using try/except because hasattr will fail to find pydantic PrivateAttrs
+                try:
+                    datasource._data_context = self._data_context
+                except ValueError:
+                    logger.debug(
+                        f'Failed to attach data context to datasource "{name}" because {datasource} has no attribute "_data_context".'
+                    )
+                    pass
```
Comment (Kilo59) [inline great_expectations/datasource/fluent/sources.py:286]: `hasattr` may fail but I believe `getattr` works. I ran into something similar yesterday.

**Review by Kilo59 — APPROVED** [review]

File: great_expectations/datasource/fluent/sources.py (hunk 3 above)
Comment (NathanFarmer) [reply great_expectations/datasource/fluent/sources.py:286]: Rolling this tip into my follow-up PR

File: great_expectations/datasource/fluent/pandas_datasource.py [hunk 4]
```diff
@@ -187,6 +205,56 @@ def _validate_batch_request(self, batch_request: BatchRequest) -> None:
...
+        exclude_fields: dict[int | str, Any] = self._include_exclude_to_dict(
+            include_exclude=exclude
+        )
+        # don't check fields that should always be set
+        check_fields: set[str] = self.__fields_set__.copy().difference(
+            _FIELDS_ALWAYS_SET
+        )
+        for field in check_fields:
+            if isinstance(getattr(self, field), tuple(_EXCLUDE_TYPES_FROM_JSON)):
+                exclude_fields[field] = True
```
Comment (NathanFarmer) [inline great_expectations/datasource/fluent/pandas_datasource.py:243]: Excluded types here

File: great_expectations/datasource/fluent/pandas_datasource.py [hunk 5]
```diff
@@ -358,6 +426,60 @@ def test_connection(self, test_assets: bool = True) -> None:
...
+            for asset_name, asset in self.assets.items():
+                # don't check fields that should always be set
+                check_fields: set[str] = asset.__fields_set__.copy().difference(
+                    _FIELDS_ALWAYS_SET
+                )
+                for field in check_fields:
+                    if isinstance(
+                        getattr(asset, field), tuple(_EXCLUDE_TYPES_FROM_JSON)
+                    ):
+                        exclude_fields["assets"] = {asset_name: {field}}
```
Comment (alexsherstinsky) [inline great_expectations/datasource/fluent/pandas_datasource.py:468]: @NathanFarmer It seems that it is possible for this line to be executed more than once; does this then mean that `exclude_fields["assets"]` will be re-initialized when this happens?  Thanks.

**Pull request merged**